File: CS/9.1/Linq/Module.cs
using DevExpress.ExpressApp;
using DevExpress.ExpressApp.NodeWrappers;
using DevExpress.ExpressApp.InfoGenerators;
using DevExpress.Xpo;

namespace Dennis.Linq {
    public sealed partial class LinqModule : ModuleBase {
        public LinqModule() {
            InitializeComponent();
        }
        public override void UpdateModel(Dictionary model) {
            base.UpdateModel(model);
            ApplicationNodeWrapper applicatioNodeWrapper = new ApplicationNodeWrapper(model);
            foreach (ClassInfoNodeWrapper classInfo in applicatioNodeWrapper.BOModel.Classes) {
                if (XafTypesInfo.Instance.FindTypeInfo(classInfo.ClassTypeInfo.Type).FindAttribute<PersistentAttribute>()!=null){
                    if (!string.IsNullOrEmpty(classInfo.Name)) {
                        ListViewInfoNodeWrapper linqViewInfo = new ListViewInfoNodeWrapper();
                        linqViewInfo.ClassName = classInfo.Name;
                        linqViewInfo.Id = ListViewNodeGenerator.GetListViewId(linqViewInfo.ClassName) + LinqCollectionSource.DefaultSuffix;
                        linqViewInfo.Node.AddChildNode(ColumnsNodeWrapper.NodeName);
                        if (applicatioNodeWrapper.Views.FindViewById(linqViewInfo.Id) == null) {
                            applicatioNodeWrapper.Views.Node.AddChildNode(linqViewInfo.Node);
                        }
                    }
                }
            }
        }
    }
}

File: VB/E859VB/Linq/Module.vb
Imports Microsoft.VisualBasic
Imports System
Imports DevExpress.ExpressApp
Imports DevExpress.ExpressApp.NodeWrappers
Imports DevExpress.ExpressApp.InfoGenerators
Imports DevExpress.Xpo

Namespace Dennis.Linq
	Public NotInheritable Partial Class LinqModule
		Inherits ModuleBase
		Public Sub New()
			InitializeComponent()
		End Sub
		Public Overrides Sub Setup(ByVal application As XafApplication)
			MyBase.Setup(application)
			AddHandler application.CreateCustomCollectionSource, AddressOf LinqCollectionSourceHelper.CreateCustomCollectionSource
		End Sub
		Public Overrides Sub UpdateModel(ByVal model As Dictionary)
			MyBase.UpdateModel(model)
			Dim applicatioNodeWrapper As New ApplicationNodeWrapper(model)
			For Each classInfo As ClassInfoNodeWrapper In applicatioNodeWrapper.BOModel.Classes
				If classInfo.ClassTypeInfo.IsPersistent Then
					If (Not String.IsNullOrEmpty(classInfo.Name)) Then
						For Each method As String In LinqCollectionSourceHelper.GetXPQueryMethods(classInfo.ClassTypeInfo.Type)
							Dim linqViewInfo As New ListViewInfoNodeWrapper()
							linqViewInfo.ClassName = classInfo.Name
							linqViewInfo.Id = ListViewNodeGenerator.GetListViewId(linqViewInfo.ClassName) & "_" & method & LinqCollectionSource.DefaultSuffix
							linqViewInfo.Node.SetAttribute("XPQueryMethod", method)
							Dim columns() As String = LinqCollectionSourceHelper.GetDisplayableProperties(classInfo.ClassTypeInfo.Type, method)
							If columns Is Nothing Then
								ListViewNodeGenerator.Generate(linqViewInfo, classInfo)
							Else
								linqViewInfo.Node.AddChildNode(ColumnsNodeWrapper.NodeName)
								For Each column As String In columns
									linqViewInfo.Columns.AddColumn(column)
								Next column
							End If
							If applicatioNodeWrapper.Views.FindViewById(linqViewInfo.Id) Is Nothing Then
								applicatioNodeWrapper.Views.Node.AddChildNode(linqViewInfo.Node)
							End If
						Next method
					End If
				End If
			Next classInfo
		End Sub
	End Class
End Namespace
